---
title: "Bioinformatic Class 18"
author: "Pallas Duong"
date: "5/31/2018"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

#**SECTION 1:** Exploring the GDC online

**Q1.** How many Cases (i.e. patient samples) have been found to have p53 mutations?

**Q2**. What are the top 6 misssense mutations found in this gene? 
HINT: Scroll down to the ‘TP53 - Protein’ section and mouse over the displayed plot. For example R175H is found in 156 cases.

**Q3.** Which domain of the protein (as annotated by PFAM) do these mutations reside in?

**Q4.** What are the top 6 primary sites (i.e. cancer locations such as Lung, Brain, etc.) with p53 mutations and how many primary sites have p53 mutations been found in? 
HINT: Clicking on the number links in the Cancer Distribution section will take you to a summary of available data accross cases, genes, and mutations for p53. Looking at the cases data will give you a ranked listing of primary sites.

**Q5.** What is the most frequentely mutated position associated with cancer in the KRas protein (i.e. the amino acid with the most mutations)?

**Q6.** Are KRas mutations common in Pancreatic Adenocarcinoma (i.e. is the Pancreas a common ‘primary site’ for KRas mutations?).

**Q7.** What is the ‘TGCA project’ with the most KRas mutations?

**Q8.** What precent of cases for this ‘TGCA project’ have KRas mutations and what precent of cases have p53 mutations? 
HINT: Placing your mouse over the project bar in the Cancer Distribution panel will bring up a tooltip with useful summary data.

**Q9.** How many TGCA Pancreatic Adenocarcinoma cases (i.e. patients from the TCGA-PAAD project) have RNA-Seq data available?

#**SECTION 2:** The GenomicDataCommons R package 
Download Package
```{r}
#source("https://bioconductor.org/biocLite.R")
#biocLite(c("GenomicDataCommons", "maftools"))
```

Load the packages
```{r}
library(GenomicDataCommons)
library(maftools)
```

Check GDC status
```{r}
GenomicDataCommons::status()
```

#**SECTION 3:** Querying the GDC from R

Using function cases() to find the numner of cases/patients acress different projects within the GDC. facet() and aggregations() act to group all cases by the project id and then count them up 
```{r}
cases_by_project <- cases() %>%
  facet("project.project_id") %>%
  aggregations()
head(cases_by_project)
```

**Q10.** Write the R code to make a barplot of the cases per project. Lets plot this data with a log scale for the y axis (log="y"), rotated axis labels (las=2) and color the bar coresponding to the TCGA-PAAD project.
```{r}
a <- cases_by_project$project.project_id
x <- a$doc_count
c <- ifelse(a$key=="TCGA-PAAD", "red", "light blue")

#cc <- rep("light blue", length(a$key != "TCGA-PAAD"))
#cc[a$key == "TCGA-PAAD"] <- "red"

#mycols <- rep("light blue", nrow(x))
#ind <- which(a[,1] =="TCGA-PAAD")
#mycols[ind]= "red"
par(mar=c(8,4,2,0)) 
barplot(x, las = 2,
        log = "y",
        col = c,
        names.arg = a$key)
```

Use another code and adapt it to answer 'find all gene expression data files for all pancreatic cancer patients'
```{r}
## This code snipet is taken from the package vignette
file_records <- files() %>%
  filter(~ cases.project.project_id == "TCGA-PAAD" &
    data_type == "Gene Expression Quantification" &
    analysis.workflow_type == "HTSeq - Counts") #%>%
  response_all()
  
```





























