---
title: "Bioinformatics Class 15"
author: "Pallas Duong"
date: "5/22/2018"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

#Section 1: Differential Expression Analysis

```{r}
#library(dplyr)
library(DESeq2)
```

Load data files
```{r}
metaFile <- "GSE37704_metadata.csv"
countFile <- "GSE37704_featurecounts.csv"
```

Import metadat and take a peak
```{r}
colData = read.csv(metaFile, row.names=1)
head(colData)
```

Import countData
```{r}
countData = read.csv(countFile, row.names=1)
head(countData)
```

Remove the first odd column in countData 
```{r}
countData <- as.matrix(countData[,-1])
head(countData)
```

Filter 0 read in countData 
```{r}
countData = countData[rowSums(countData)>1, ]
head(countData)
```

Set up the DESeqDataSet object and run the DESeq pipeline
```{r}
dds = DESeqDataSetFromMatrix(countData=countData,
                             colData=colData,
                             design=~condition)
dds = DESeq(dds)
```

```{r}
dds
```

How many genes?
```{r}
nrow(countData)
```

HoxA1 knockdown vs. control siRNA 
```{r}
res = results(dds, contrast=c("condition", "hoxa1_kd", "control_sirna"))
```

Reorder these reulsts by p-value and call summary() on the results objects to get a sense of how many genes are up or down-regulated at the default FDR of 0.1
```{r}
res = res[order(res$pvalue),]
summary(res)
```

Pathway analysis downstream will use KEGG pathways, and genes in KEGG pathways are annotated with Entrez gene IDs. Add them.
```{r}
library("AnnotationDbi")
library("org.Hs.eg.db")

columns(org.Hs.eg.db)
```

```{r}
res$symbol = mapIds(org.Hs.eg.db,
                    keys=row.names(res), 
                    column="SYMBOL",
                    keytype="ENSEMBL",
                    multiVals="first")

res$entrez = mapIds(org.Hs.eg.db,
                    keys=row.names(res), 
                    column="ENTREZID",
                    keytype="ENSEMBL",
                    multiVals="first")

res$name =   mapIds(org.Hs.eg.db,
                    keys=row.names(res), 
                    column="GENENAME",
                    keytype="ENSEMBL",
                    multiVals="first")

head(res, 10)
```

#Section 2: Pathway Analysis 

Install required bioconductor packages 
```{r}
source("http://bioconductor.org/biocLite.R")
biocLite( c("pathview", "gage", "gageData") )
```

Load the packages and set up the KEGG data-sets
```{r}
library(pathview)
library(gage)
library(gageData)

data(kegg.sets.hs)
data(sigmet.idx.hs)

kegg.sets.hs = kegg.sets.hs[sigmet.idx.hs]
head(kegg.sets.hs, 3)
```

Name vector of fold changes, where the names of the values are the Entriz gene IDs
```{r}
foldchanges = res$log2FoldChange
names(foldchanges) = res$entrez
head(foldchanges)
```

Get the reults 
```{r}
keggres = gage(foldchanges, gsets=kegg.sets.hs, same.dir=TRUE)
```

Look at the result object
```{r}
attributes(keggres)
```

```{r}
head(keggres$greater)
```

```{r}
head(keggres$less)
```

Look at both up an down and statistics 
```{r}
lapply(keggres, head)

```

Process the results to pull out the top 5 upregulated pathways, then further process that just to get the IDs

```{r}
pathways = data.frame(id=rownames(keggres$greater), keggres$greater)
head(pathways)
```

Make a pathway plot with the result shown in color
```{r}
pathview(gene.data=foldchanges, pathway.id="hsa04110")
```
![](hsa04110.pathview.png)

Change the display: generating a PDF graph
```{r}
pathview(gene.data=foldchanges, pathway.id="hsa04110", kegg.native=FALSE)

```

Focus on top 5 upregulated pathways
```{r}
keggrespathways <- rownames(keggres$greater)[1:5]
```

Extract the IDs part of each string 
```{r}
keggresids = substr(keggrespathways, start=1, stop=8)
keggresids
```

Draw plots for all the top 5 pathwyas
```{r}
pathview(gene.data=foldchanges, pathway.id=keggresids, species="hsa")

```
![](hsa04640.pathview.png)
![](hsa04630.pathview.png)
![](hsa04142.pathview.png)
![](hsa00140.pathview.png)
![](hsa04740.pathview.png)

#Section 3: Gene Ontology (GO)

Biological Process
```{r}
data(go.sets.hs)
data(go.subs.hs)
gobpsets = go.sets.hs[go.subs.hs$BP]

gobpres = gage(foldchanges, gsets=gobpsets, same.dir=TRUE)

lapply(gobpres, head)
```
 
#Section 4: Reactome Pathway Analysis 

Over-representation enrichment analysis and pathway-topology analysis with Reactome 
```{r}
sig_genes <- res[res$padj <= 0.05 & !is.na(res$padj), "symbol"]
print(paste("Total number of significant genes:", length(sig_genes)))
```

```{r}
write.table(sig_genes, file="significant_genes.txt", row.names=FALSE, col.names=FALSE, quote=FALSE)

```

Perform pathway analysis online using Reactome website (https://reactome.org/PathwayBrowser/#TOOL=AT)

**Q** The pathway has the most significant "Entities p-value" is Endosomal/Vascuolar pathway. The most significant pathways listed do not match with my previous KEGG results

#Section 5: GO Analysis

Perform Gene Set GO Enrichlment online using (http://www.geneontology.org/page/go-enrichment-analysis)










